url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/beans/XMLDecoder.java

```diff
@@ -0,0 +1,705 @@
+                    boolean isAssignable = isNull? false : parameterTypes[i].isAssignableFrom(clazzes[i]);
+                    if ( isNull || isPrimitive || isAssignable ) {
+                        continue;
+                    }
+                    match = false;
+                }
+                if (match) {
+                    matchMethods.add(method);
+                }
+            }
+
+            int size = matchMethods.size();
+            if (size == 1) {
+                // Only one method matches, just invoke it
+                return matchMethods.get(0);
+            } else if (size == 0) {
+                // Does not find any matching one, throw exception
+                throw new NoSuchMethodException(Messages.getString(
+                        "custom.beans.41", methodName)); //$NON-NLS-1$
+            }
+
+            // There are more than one method matching the signature
+            // Find the most specific one to invoke
+            MethodComparator comparator = new MethodComparator(methodName,
+                    clazzes);
+            Method chosenOne = matchMethods.get(0);
+            matchMethods.remove(0);
+            int methodCounter = 1;
+            for (Method method : matchMethods) {
+                int difference = comparator.compare(chosenOne, method);
+                if (difference > 0) {
+                    chosenOne = method;
+                    methodCounter = 1;
+                } else if (difference == 0) {
+                    methodCounter++;
+                }
+            }
+            if (methodCounter > 1) {
+                // if 2 methods have same relevance, throw exception
+                throw new NoSuchMethodException(Messages.getString(
+                                                                   "custom.beans.62", methodName)); //$NON-NLS-1$
+            }
+            return chosenOne;
+        }
+
+        private boolean isPrimitiveWrapper(Class<?> wrapper, Class<?> base) {
+            return (base == boolean.class) && (wrapper == Boolean.class)
+                    || (base == byte.class) && (wrapper == Byte.class)
+                    || (base == char.class) && (wrapper == Character.class)
+                    || (base == short.class) && (wrapper == Short.class)
+                    || (base == int.class) && (wrapper == Integer.class)
+                    || (base == long.class) && (wrapper == Long.class)
+                    || (base == float.class) && (wrapper == Float.class)
+                    || (base == double.class) && (wrapper == Double.class);
+        }
+
+        private String capitalize(String str) {
+            StringBuilder buf = new StringBuilder(str);
+            buf.setCharAt(0, Character.toUpperCase(buf.charAt(0)));
+            return buf.toString();
+        }
+
+        @SuppressWarnings("nls")
+        private Object executeBasic(Elem elem) throws Exception {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.